app/gui/view/graph-editor/src/lib.rs

```diff
@@ -3652,3 +3649,170 @@ impl display::Object for GraphEditor {
         self.model.display_object()
     }
 }
+
+
+
+// =============
+// === Tests ===
+// =============
+
+#[cfg(test)]
+mod graph_editor_tests {
+    use super::*;
+
+    #[test]
+    fn test_adding_node_by_internal_api() {
+        let (_, graph_editor) = init();
+        assert_eq!(graph_editor.nodes().len(), 0);
+        graph_editor.add_node();
+        assert_eq!(graph_editor.nodes().len(), 1);
+        graph_editor.assert(Case { node_source: None, should_edit: false });
+    }
+
+    #[test]
+    fn test_adding_node_by_shortcut() {
+        let add_node = |editor: &GraphEditor| editor.start_node_creation();
+        test_adding_node(add_node);
+    }
+
+    #[test]
+    fn test_adding_node_by_adding_node_button() {
+        let add_node = |editor: &GraphEditor| {
+            let adding_node_button = &editor.model.add_node_button;
+            adding_node_button.click();
+        };
+        test_adding_node(add_node);
+    }
+
+    fn test_adding_node(add_node: impl Fn(&GraphEditor)) {
+        let (app, graph_editor) = init();
+        assert_eq!(graph_editor.nodes().len(), 0);
+
+        // Adding first node.
+        let (node_1_id, node_1) = graph_editor.add_node_by(&add_node);
+        graph_editor.assert(Case { node_source: None, should_edit: true });
+        graph_editor.stop_editing();
+        assert_eq!(graph_editor.nodes().len(), 1);
+
+        // First node is created in the center of the screen.
+        let node_1_pos = node_1.position();
+        let screen_center = app.display.default_scene.screen_to_scene_coordinates(Vector3::zeros());
+        assert_eq!(node_1_pos.xy(), screen_center.xy());
+        graph_editor.model.nodes.select(node_1_id);
+
+        // Adding second node.
+        let (_, node_2) = graph_editor.add_node_by(&add_node);
+        graph_editor.assert(Case { node_source: Some(node_1_id), should_edit: true });
+        assert_eq!(graph_editor.nodes().len(), 2);
+
+        // Second node is below the first.
+        let node_2_pos = node_2.position();
+        assert!(node_2_pos.y < node_1_pos.y);
+    }
+
+    #[test]
+    fn test_adding_node_by_dropping_edge() {
+        let (app, graph_editor) = init();
+        assert_eq!(graph_editor.model.nodes.all.len(), 0);
+        // Adding a new node.
+        let (node_1_id, node_1) = graph_editor.add_node_by_api();
+        graph_editor.stop_editing();
+        // Creating edge.
+        let port = node_1.model.output_port().expect("No output port");
+        port.events.mouse_down.emit(());
+        assert_eq!(graph_editor.edges().len(), 1);
+        //node_1.model.output.test_port_press();
+        // Dropping edge.
+        let mouse = &app.display.default_scene.mouse;
+        let click_pos = Vector2(300.0, 300.0);
+        mouse.frp.position.emit(click_pos);
+        let click_on_background = |_: &GraphEditor| mouse.click_on_background();
```

Comment:
Now `click_on_background` emits both `up` and `down`. It was working before because of specific implementation details of GraphEditor, so it makes sense, thank you.